core/src/main/java/org/opensearch/sql/expression/function/udf/condition/EnhancedCoalesceFunction.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.expression.function.udf.condition;
+
+import java.util.List;
+import java.util.function.Supplier;
+import org.apache.calcite.adapter.enumerable.NotNullImplementor;
+import org.apache.calcite.adapter.enumerable.NullPolicy;
+import org.apache.calcite.linq4j.tree.Expression;
+import org.apache.calcite.linq4j.tree.Expressions;
+import org.apache.calcite.sql.type.SqlReturnTypeInference;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.opensearch.sql.data.model.ExprValue;
+import org.opensearch.sql.data.model.ExprValueUtils;
+import org.opensearch.sql.expression.function.ImplementorUDF;
+import org.opensearch.sql.expression.function.UDFOperandMetadata;
+
+public class EnhancedCoalesceFunction extends ImplementorUDF {
+
+  public EnhancedCoalesceFunction() {
+    super(createImplementor(), NullPolicy.NONE);
+  }
+
+  private static NotNullImplementor createImplementor() {
+    return (translator, call, translatedOperands) -> {
+      List<Expression> exprValues =
+          translatedOperands.stream()
+              .map(
+                  operand ->
+                      (Expression)
+                          Expressions.call(
+                              ExprValueUtils.class,
+                              "fromObjectValue",
+                              Expressions.convert_(operand, Object.class)))
+              .toList();
+
+      Expression returnTypeName = Expressions.constant(call.getType().getSqlTypeName().toString());
```

Comment:
"the least restrictive common type among all arguments", this is determined by Calcite?